fdbclient/BlobConnectionProvider.cpp

```diff
@@ -44,18 +50,42 @@ struct PartitionedBlobConnectionProvider : BlobConnectionProvider {
 		return std::pair(conn, metadata.partitions[writePartition].toString() + newFileName);
 	}
 
-	Reference<BackupContainerFileSystem> getForRead(std::string filePath) { return conn; }
+	Reference<BackupContainerFileSystem> getForRead(std::string filePath) {
+		CODE_PROBE(isExpired(), "partitioned blob connection using expired blob metadata for read!");
+		return conn;
+	}
 
-	PartitionedBlobConnectionProvider(const Standalone<BlobMetadataDetailsRef> metadata) : metadata(metadata) {
-		ASSERT(metadata.base.present());
-		ASSERT(metadata.partitions.size() >= 2);
-		conn = BackupContainerFileSystem::openContainerFS(metadata.base.get().toString(), {}, {});
-		for (auto& it : metadata.partitions) {
+	void updateMetadata(const Standalone<BlobMetadataDetailsRef>& newMetadata, bool checkPrevious) {
+		ASSERT(newMetadata.base.present());
+		ASSERT(newMetadata.partitions.size() >= 2);
+		for (auto& it : newMetadata.partitions) {
 			// these should be suffixes, not whole blob urls
 			ASSERT(it.toString().find("://") == std::string::npos);
 		}
+		if (checkPrevious) {
```

Comment:
not sure if I understand the goal of this loop? Also, do we only need to worry about `expireAt` and not `refreshAt`?

Reply:
The point of this is to validate that the storage location didn't change for the same blob connection between refreshes. The only reason we do refreshes is because credentials expire.

Reply:
The check of expireAt is for idempotency and ordering. The assumption is that expireAt is always (original request time + interval) and that refreshAt is just a smaller interval after original request time.

We could assert that in this case that newMetadata.refreshAt > metadata.refreshAt, if  newMetadata.expireAt > metadata.expireAt, but it doesn't seem like a very useful requirement.

Reply:
> The assumption is that expireAt is always (original request time + interval) and that refreshAt is just a smaller interval after original request time.

It is possible for refreshAt = -1 (never refresh), but expireAt = validInterval, essentially it would mean that KMS wants the credentials to be only used for a defined interval, hence, never refresh

Reply:
Ok, that makes sense. That still means that expireAt is effectively the source of truth for which credential is newer for the same resource, correct?
Also IMO we should use std::numeric_limits<max> for never refresh instead of -1, which is what I did in a couple places here, as then the logic of comparing now() to the refresh at time is always correct

Reply:
> Also IMO we should use std::numeric_limits<T>::max()

Given the value is set external to FDB, I kept it as `-1` as the external service implementations may have different interpretations for max etc (unlikely); but sure can be done.

Reply:
Makes sense, IMO as long as we're just consistent in however FDB represents it, then it's up to the KmsConn* implementation(s) to translate properly from whatever external KMS has to fdb's semantics

Reply:
Sounds reasonable, I like `numeric_limit` we can standardize on it and will update KMSConn layer to do the proper translation. 
Thanks